handle tags and tags_exceptions in multi-tag articles

nhohung opened the pull request:
Hi Marcin @mmagnus,

### Issue with multi-tag articles

Pocket articles may have multiple associated tags. As mentioned in #33, an user may organize his/her Pocket list by year and month tags, for example 2 tags `2020` and `month05` in this order for one article. The tag attribute of this article will look something like this:
`'tags': {'2020': {'item_id': '3325357131', 'tag': '2020'}, 'month05': {'item_id': '3325357131', 'tag': 'month05'}}`

When defining `TAGS` in the plugin setting, only the first tag is checked (ie. `2020`). A tag filter of only 2020 articles will work (`TAGS = ['2020']`). However if an user wants to filter articles in May, `TAGS = ['month05']` will return nothing because the second tag is not read in the current version. Running this plugin gives `No articles in the Pocket account to download`.

### This fix

In this branch, I made three small changes:
- extract **all tags** instead of just the first one at line 307
- skip articles with **any tag** in `TAGS_EXCEPTIONS` at line 315
- keep articles with **any tag** in `TAGS` from lines 326-331 (after filtering `TAGS_EXCEPTIONS`)

### Comments

I have tested different combinations of `TAGS` and `TAGS_EXCEPTIONS` (in my case, filtering by either year, month or both). The branch works as expected.

#### Update 1 (15/05/2021):
@alvaroreig pointed out 3 bugs in this branch. They're fixed in the follow up commits (to be tested), see comments below:
- multiple tags per article functionality doesn't work
- autotags functionality crashes
- INCLUDE_UNTAGGED doesn't work

#### Update 2 (22/05/2021):
I have added the latest commit to fix #33 on top of the 3 reported issues in Update 1.

Can you please try it and let me know if any change is needed?

Cheers,
Hung

alvaroreig commented:
Hi there @nhohung @mmagnus 

I ran a quick test and I don't think the fix works as expected. 

First, the multiple tag per article functionality doesn't seem to work. With this configuration:

TAGS = ['tag1','tag2']

and this articles:

article1 => tag1,tag2
article2 => tag1

I only got

tag1 => article1, article2

While I was expecting a tag2 section as well

Second of all, this fix broke the autotags functionality. If I use this configuration:

TAGS = []

The plugin crashes:


>   TypeError: unhashable type: 'list'
>   Error in 
> 
>  sys.excepthook:
> 
>   Traceback (most recent call last):
>     File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 153, in apport_excepthook
>       with os.fdopen(os.open(pr_filename,
>   FileNotFoundError: [Errno 2] No existe el archivo o el directorio: '/var/crash/_usr_bin_ebook-convert.1000.crash'
>   
>   Original exception was:
>   Traceback (most recent call last):
>     File "/usr/bin/ebook-convert", line 20, in <module>
>       sys.exit(main())
>     File "/usr/lib/calibre/calibre/ebooks/conversion/cli.py", line 401, in main
>       plumber.run()
>     File "/usr/lib/calibre/calibre/ebooks/conversion/plumber.py", line 1108, in run
>       self.oeb = self.input_plugin(stream, self.opts,
>     File "/usr/lib/calibre/calibre/customize/conversion.py", line 245, in __call__
>       ret = self.convert(stream, options, file_ext,
>     File "/usr/lib/calibre/calibre/ebooks/conversion/plugins/recipe_input.py", line 139, in convert
>       ro.download()
>     File "/usr/lib/calibre/calibre/web/feeds/news.py", line 1022, in download
>       res = self.build_index()
>     File "/usr/lib/calibre/calibre/web/feeds/news.py", line 1190, in build_index
>       feeds = feeds_from_index(self.parse_index(), oldest_article=self.oldest_article,
>     File "<string>", line 320, in parse_index
>   TypeError: unhashable type: 'list'

It makes sense, because in the autotags section (line 320) I can't see any reference to the new tagcheck variable.

And third, INCLUDE_UNTAGGED doesn't work in this branch. I don't get any untagged articles at all. However, it does work for me in the master branch.

I may have tested it wrong, as I couldn't find the time to propperly run regression tests. But it doesn't look good I think.

Tested in linux with python 3.8.5

Regards,

nhohung commented:
Hi @alvaroreig,

Thank you for testing it. I've just added a new commit to address your second point (autotags with `TAGS=[]`). Feel free to give it a go. Feedback is appreciated.

I'll look at the other 2 points later.

Cheers,
Hung

alvaroreig commented:
Hi there @nhohung ,

I tested your last commit and AUTOTAGS no longer crashes, but it still didn't work as I was expecting. I got the same results as with explicit tags:

> TAGS = ['tag1','tag2'] OR
> TAGS = [] 
> 
> and this articles:
> 
> article1 => tag1,tag2
> article2 => tag1
> 
> I only got
> 
> tag1 => article1, article2
> 
> While I was expecting a tag2 section as well

I found that weird, considering that on first glance the code seem correct. And then I remember something @mmagnus  and I talked some months ago. The recipe, by default, doesnt allow duplicates. The setting is on line 176:

`        ignore_duplicate_articles = {'url'}`

Thats why my test case doesn't work. article1 is already in tag1 section, and therefore tag2 section is empty. If I remove that setting:

`        ignore_duplicate_articles = {}`

Then I got the expected result:

> tag1 => article1, article2
> tag2 => article1

I think that, therefore, we sould set a new flag ALLOW_DUPLICATES that is proyected to line 176, as this new functionality "consider multiple tags in every article" might work weird or not at all if ALLOW_DUPLICATES=False. @mmagnus , what do you think?

I didn't test the TAG_EXCEPTIONS functionality yet.

Regards,

nhohung commented:
Hi @alvaroreig,

I have added three separate commits today to address your remaining points:
- Untagged articles: I have edited line 310 to convert `'Untagged'` into a list that can be accessed by `tagcheck in tag`. `INCLUDE_UNTAGGED` option should work now, please try it.
- Duplicate tags between articles: I see your point in the comment yesterday and have put a new flag `ALLOW_DUPLICATES` in line 58. This flag controls `ignore_duplicate_articles` in lines 179-182. 
- Last commit adds description for the new flag `ALLOW_DUPLICATE`.

I created some dummy articles and the fix runs as expected. Please test and see if it works for you.

Thanks,
Hung

nhohung commented:
In the latest commit I've added an error handling for `status=2` articles at . This fixes issue #33.

nhohung commented:
In the latest commit, I filter `status=2` articles from `response` dictionary upfront instead of filtering it by a `KeyError` exception when extracting article information. Particularly:
- Removed `KeyError` exception in the loop at line 347 (this can be left without any issue but I just remove it to avoid garbage collection and to stay as close with the original code as possible).
- Added a condition check at line 299: if an article has `status=2`, delete that article from the dictionary. For this to work, I need to make a copy of the dictionary (line 298) because dictionary size can't be changed while iteration.

This should fix issue #38.
